web/ui/react-app/src/api/Fetch.tsx

```diff
@@ -0,0 +1,75 @@
+import { Component } from 'react';
+import { fetchAPI } from './FetchAPI';
+
+export interface FetchState<T> {
+  data?: T;
+  error?: Error;
+}
+
+interface FetchProps {
+  url?: string;
+  urls?: string[];
+  options?: RequestInit;
+  children: (state: any) => any;
+}
+
+const urlsEqual = (urls?: string[], nextURLs?: string[]) => {
+  if (!urls || !nextURLs) {
+    return false;
+  }
+  return JSON.stringify(urls) === JSON.stringify(nextURLs);
+};
+
+export class Fetch extends Component<FetchProps, FetchState<any>> {
+  state = {
+    data: undefined,
+    error: undefined,
+  };
+
+  componentDidMount() {
+    const { url, urls, options } = this.props;
+    if (Boolean(url) && Boolean(urls)) {
+      throw new Error('Please provide either url or urls param but not both.');
+    }
+    if (urls && urls.length) {
+      this.handleResponse(() => this.getAll(urls, options));
+    } else if (url) {
+      this.handleResponse(() => fetchAPI(url, options).then(res => res.data));
+    } else {
+      throw new Error('URL/s is Missing');
+    }
+  }
+
+  componentDidUpdate(nextProps: FetchProps) {
+    const { url: nextURL, urls: nextURLs, options } = nextProps;
+    const { url, urls } = this.props;
+    if (nextURLs && !urlsEqual(urls, nextURLs)) {
+      this.handleResponse(() => this.getAll(nextURLs, options));
+    } else if (nextURL && url !== nextURL) {
+      this.handleResponse(() => fetchAPI(nextURL, options).then(res => res.data));
+    }
+  }
+
+  getAll = async (urls: string[], options?: RequestInit) => {
+    return await Promise.all(
+      urls
+        .map(async url => {
+          const result = await fetchAPI(url, options);
+          return result.data;
+        })
+        .filter(Boolean)
+    );
+  };
+
+  handleResponse = async (resultCb: () => Promise<any>) => {
+    try {
+      this.setState({ data: await resultCb() });
+    } catch (error) {
+      this.setState({ error });
+    }
+  };
+
+  render() {
+    return this.props.children(this.state);
+  }
+}
```

Comment:
why not use hooks in this component? the component lifecycle methods you have are the ones a hook can replace

Reply:
In hooks you really don't have a mechanism to compare dependency array values (like ```this.props.foo === nextProps.foo```) which is needed when you expect new instances every time but not changes in the data

Reply:
How often do you expect these urls to change? I guess I'm having a hard time understanding how I would use this mechanism

Reply:
Actually i think i understand now... eg if params for a fetch change

Reply:
hooks return a setState for managing the url, which can be used to re-trigger an effect

Reply:
The issue is not with the dependencies updates (url, options). The issue is when the Effect is triggered without those values to be changed and this is happening when on every render new instance is passed of those values. To overcome this problem you have to implement logic similar to the one which class components gives you anyway.

Reply:
Any ideas on how I can improve this? Tried @hooten suggestion but I'm a bit unhappy with that the user is forced to pass array for a single request. On the other side, @hooten is right that a single prop will be less confusing. I can just define it as ```endpoint: string | string[]``` but this will turn into the same situation because of which we split useFetches into two hooks. WDYT? @juliusv @csmarchbanks

Reply:
@boyskila Was the last comment meant to be posted on the other sub-thread https://github.com/prometheus/prometheus/pull/6267#discussion_r342867962?

Both options (always passing an array vs. having two ways of passing URLs) seem equally bad to me, but I don't see a better way, unless we get rid of the multi-fetch functionality completely. So maybe we just have to live with the ugliness unless someone has another design idea.

Reply:
Oh, my bad. It is for https://github.com/prometheus/prometheus/pull/6267#discussion_r342867962

Reply:
I think I agree with @juliusv that both options are less than ideal. Maybe it would be better to remove the multi-fetch functionality. It might be an anti-pattern to wait for multiple remote calls to return before beginning rendering.

@csmarchbanks do you have any thought on this (since you wrote the previous fetch code)?

Reply:
Other option could be to create a new route in the api.go to handle all the information at once
``` /stats ``` and thus we will not need multi fetch anymore

Reply:
I have been giving the multi-fetch functionality some thought over the last few days and generally think it should go away for the following reasons.
1. IMO Lumping all requests together and returning an error if any of them errors is an anti-pattern. Errors from each call should be handled separately with as much information exposed to the user as possible from any calls that do succeed.
1. The function is complicated (I am not convinced there aren't unexpected edge cases that will bite us) and only used in one spot. Doing three `useFetch` calls is not much more code, and will be much more readable in the future. No need for a new route either.
1. What about passing extra options such as if doing multiple POST requests? Do we now need to support arrays of both urls and options? Or some wrapping object?

At that point, this component becomes of dubious value, perhaps FetchWithStatus is still useful to reduce a little bit of code. The Go developer inside of me still wants to make the page explicitly handle errors though.

Reply:
@csmarchbanks Thank you for your feedback!

For point 1 I would say that it depends on the requirements.
In the Status page, I want to show the error message if some of the 3 requests fail but probably not in such a way. You are right that it is good to pass all the information back to the user and the user to decide how to handle the errors. So, for example, if we fetch ```[A, B, C]``` and the response is ```[success, fail, success]``` is good to return back the response objects instead of the data only.

For point 2:  This is a basic implementation covering the needs of the app. I'm sure that I'm far from what it should be to cover all edge cases and scenarios. Not sure how the three useFetches will handle the requirements for the status page?

For point 3: Currently, you can pass only one ```RequestInit``` object which will not work for multiple requests as you mention but I guess it can be easily implemented just by sending not ```string[]``` but ```{url: string, opt: RequestInit}[]```

Can you please explain why do you think the function is complicated? IMO, at least for now, is not doing so much  :)

Reply:
Here is how the status page could use three fetches to accomplish exactly what `useFetches` does today: https://github.com/csmarchbanks/prometheus/commit/a5fd258404520e51c3664631153c183490426e0f

As far as complexity goes: 
Currently in master the `useFetches` function is not too complicated - but also requires the hacky workaround for dependencies. In this PR you have to create an entire component just to fetch some data which seems like a lot, especially for components that have a different lifecycle than just a page load.

Reply:
Thanks for the suggestion, I like the approach.